File: day71-80/day78/1106. Parsing A Boolean Expression.js
class Solution {
  parseBoolExpr(expression) {
    const parse = (expression) => {
      if (expression === "t") {
        return true;
      } else if (expression === "f") {
        return false;
      }

      if (expression[0] === "!") {
        return parseNot(expression.slice(2, -1));
      }

      if (expression[0] === "&") {
        return parseAnd(expression.slice(2, -1));
      }

      if (expression[0] === "|") {
        return parseOr(expression.slice(2, -1));
      }

      throw new Error("Expresión inválida");
    };

    const parseNot = (subExpr) => {
      return !parse(subExpr);
    };

    const parseAnd = (subExpr) => {
      const subExpressions = splitExpressions(subExpr);
      return subExpressions.every((expr) => parse(expr));
    };

    const parseOr = (subExpr) => {
      const subExpressions = splitExpressions(subExpr);
      return subExpressions.some((expr) => parse(expr));
    };

    const splitExpressions = (expression) => {
      const subExpressions = [];
      let balance = 0;
      let currentExpr = [];

      for (const char of expression) {
        if (char === "," && balance === 0) {
          subExpressions.push(currentExpr.join(""));
          currentExpr = [];
        } else {
          currentExpr.push(char);
          if (char === "(") {
            balance += 1;
          } else if (char === ")") {
            balance -= 1;
          }
        }
      }
      subExpressions.push(currentExpr.join(""));
      return subExpressions;
    };
    return parse(expression);
  }
}

sol = new Solution();
expression = "!(&(f,t))";
console.log(sol.parseBoolExpr(expression));

File: day71-80/day78/1106. Parsing A Boolean Expression.py
class Solution:
    def parseBoolExpr(self, expression: str) -> bool:
        def parse(expression: str) -> bool:
            if expression == 't':
                return True
            elif expression == 'f':
                return False

            if expression[0] == '!':
                return parse_not(expression[2:-1])  

            if expression[0] == '&':
                return parse_and(expression[2:-1])  

            if expression[0] == '|':
                return parse_or(expression[2:-1])  

            raise ValueError("Expresión inválida")
        
        def parse_not(subExpr: str) -> bool:  
            return not parse(subExpr)

        def parse_and(subExpr: str) -> bool:   
            sub_expressions = split_expressions(subExpr)
            return all(parse(expr) for expr in sub_expressions)

        def parse_or(subExpr: str) -> bool:
            sub_expressions = split_expressions(subExpr)
            return any(parse(expr) for expr in sub_expressions)

        def split_expressions(expression: str) -> list:
            sub_expressions = []
            balance = 0
            current_expr = []
            for char in expression:
                if char == ',' and balance == 0:
                    sub_expressions.append(''.join(current_expr))
                    current_expr = []
                else:
                    current_expr.append(char)
                    if char == '(':
                        balance += 1
                    elif char == ')':
                        balance -= 1
            sub_expressions.append(''.join(current_expr))  
            return sub_expressions
        return parse(expression)


sol=Solution()
expression="!(&(f,t))"
print(sol.parseBoolExpr(expression))